### ISSUE TITLE
Naming convention not applied to tables in ASP.NET Identity

### ISSUE THREAD

**Issue by kanonlemon** [efcore/EFCore.NamingConventions#2]
.net3.0 core
in my code
```c#
options.UseNpgsql(
    Configuration.GetConnectionString("DefaultConnection")
)
.UseSnakeCaseNamingConventions()
```

the result of Migrations shows:

all columns have changed to snaked mode
but table names keep the CamelCase

**Comment by roji** [comment]
Apologies, I was accidentally not subscribed to notifications on this repo.

I cannot reproduce this issue, can you please submit precise steps which reproduce the problem? Here's what I did:

Starting program:

```c#
class Program
{
    static async Task Main(string[] args)
    {
        using var ctx = new BlogContext();
        ctx.Database.EnsureDeleted();
        ctx.Database.EnsureCreated();
    }
}

public class BlogContext : DbContext
{
    public DbSet<Blog> Blogs { get; set; }

    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
    {
        optionsBuilder.UseNpgsql("...");
    }
}

public class Blog
{
    public int Id { get; set; }
    public string Name { get; set; }
}
```

I created a migration, then added `.UseSnakeCaseNamingConvention()` in OnConfiguring, and created a second migration. At this point, when generating a database with the migrations (`dotnet ef database drop -f; dotnet ef database update`) my database was created with lower-case table names as expected.

**Comment by pawmot** [comment]
Hi @roji, I don't know if @kanonlemon encountered exactly the same issue as me, but in my case the Identity Core tables were problematic - columns were snake_case, but table names remained CamelCase. I was able to fix this by configuring the model - using just `UseSnakeCaseNamingConventions` was not enough. When it comes to my own types table names and columns were OK, but e.g. primary keys were still beginning with "PK" instead of "pk".

**Comment by jasonwarford** [comment]
I'm experiencing the same thing. In my case I'm using dotnet core 3.1 and trying to generate the migrations for Identity Server 4 PersistedGrantDbContext and ConfigurationDbContext.

I'll try to pair it down to a simple example that illustrates the issue.

**Comment by roji** [comment]
> I'll try to pair it down to a simple example that illustrates the issue.

That would be great.

**Comment by roji** [comment]
tl;dr for now, this plugin won't affect ASP.NET Identity tables - you'll have to apply naming manually (e.g. [see this SO](https://stackoverflow.com/questions/59286100/aspnetcore-identity-using-postgresql-how-to-create-structure)).

The problem is specific to the ASP.NET Identity model. Because the model uses generic types (TUser, TRole...), it [explicitly configures table names in OnModelCreating](https://github.com/dotnet/aspnetcore/blob/6e35229b015f0164c0876e2f6dc051409b1d1390/src/Identity/EntityFrameworkCore/src/IdentityDbContext.cs#L132). Anything that happens in OnModelCreating takes precedence over conventions, which is what this plugin sets up to apply the naming. In other words, the mechanism of specifying table names in OnModelConfiguring was originally meant to allow users to *override* conventions, but in this case is being used by ASP.NET Identity in another way.

@ajcvickers @AndriySvyryd any ideas here? Identity could have its own convention to set the names, but then we'd have two competing conventions with no clear ordering...

**Comment by ajcvickers** [comment]
@roji The ASP.NET Core Identity model is heavily configured--very few conventions are used in its configuration. It's seems pretty complicated to try to go back to using some conventions after this--it's not something I think you can do now. Even if we did, it would be work and a breaking change for Identity to opt into it, which I don't think will happen.

**Comment by roji** [comment]
That's what I thought. I'll close this for now...

**Comment by roji** [comment]
Note to anyone hitting this: although the plugin can't apply its naming conventions, you can do this yourself by dropping the following into your OnModelCreating:

```c#
modelBuilder.Entity<IdentityUser>().ToTable("asp_net_users");
modelBuilder.Entity<IdentityUserToken<string>>().ToTable("asp_net_user_tokens");
modelBuilder.Entity<IdentityUserLogin<string>>().ToTable("asp_net_user_logins");
modelBuilder.Entity<IdentityUserClaim<string>>().ToTable("asp_net_user_claims");
modelBuilder.Entity<IdentityRole>().ToTable("asp_net_roles");
modelBuilder.Entity<IdentityUserRole<string>>().ToTable("asp_net_user_roles");
modelBuilder.Entity<IdentityRoleClaim<string>>().ToTable("asp_net_role_claims");
```

**Comment by 10GeekJames** [comment]
to make the above compile I needed to add .HasNoKey() to each line.  This did help with those tables, thank you.  However; I seem to have a different situation where NO tables are getting the lower-case treatment.  I do have the project broken across several projects with data, infrastructure, front-end.  The front end is where something like dotnet ef database migration would be called.  When I implement this code I get lowercase "properties" across the board and zero lowercase table names, any guesses?

**Comment by roji** [comment]
You should definitely not be adding HasNoKey() to the above - are you sure you're using `string` as your key? If not, you need to change the generic type parameter above. Otherwise please post your identity DbContext.

Re the 2nd problem - are you saying you're seeing a non-Identity table not being affected by this plugin? If so, can you please open a new issue with a minimal code sample?

**Comment by igorgomeslima** [comment]
> Note to anyone hitting this: although the plugin can't apply its naming conventions, you can do this yourself by dropping the following into your OnModelCreating:
> 
> ```cs
> modelBuilder.Entity<IdentityUser>().ToTable("asp_net_users");
> modelBuilder.Entity<IdentityUserToken<string>>().ToTable("asp_net_user_tokens");
> modelBuilder.Entity<IdentityUserLogin<string>>().ToTable("asp_net_user_logins");
> modelBuilder.Entity<IdentityUserClaim<string>>().ToTable("asp_net_user_claims");
> modelBuilder.Entity<IdentityRole>().ToTable("asp_net_roles");
> modelBuilder.Entity<IdentityUserRole<string>>().ToTable("asp_net_user_roles");
> modelBuilder.Entity<IdentityRoleClaim<string>>().ToTable("asp_net_role_claims");
> ```
> 
> (don't forget to change `string` if you're using another key type)

This not work anymore.

With:

    ```
<PackageReference Include="EFCore.NamingConventions" Version="5.0.0-rc1" />
    <PackageReference Include="Microsoft.AspNetCore.Identity.EntityFrameworkCore" Version="5.0.0" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.Tools" Version="5.0.0">
    <PackageReference Include="Microsoft.Extensions.Identity.Core" Version="5.0.0" />
    <PackageReference Include="Npgsql.EntityFrameworkCore.PostgreSQL" Version="5.0.0" />
    <PackageReference Include="Npgsql.EntityFrameworkCore.PostgreSQL.Design" Version="1.1.0" />
```

**Comment by roji** [comment]
@igorgomeslima when you do ToTable, you're explicitly setting table names - EFCore.NamingConventions has nothing to do with anything at that point... Can you please double-check that there's actually a behavioral difference between 5.0.0 and a previous version?

**Comment by igorgomeslima** [comment]
> @igorgomeslima when you do ToTable, you're explicitly setting table names - EFCore.NamingConventions has nothing to do with anything at that point... Can you please double-check that there's actually a behavioral difference between 5.0.0 and a previous version?

@roji With previous versions _5.0.0-preview8 & 5.0.0-preview5_ i got the same result. With versions below those, I got the error mentioned [here](https://github.com/efcore/EFCore.NamingConventions/issues/30).

**Comment by roji** [comment]
My point is that whatever is happening with ToTable is unrelated to this plugin - you should be able to remove EFCore.NamingConventions and see exactly the same behavior as you see with it.

**Comment by igorgomeslima** [comment]
> My point is that whatever is happening with ToTable is unrelated to this plugin - you should be able to remove EFCore.NamingConventions and see exactly the same behavior as you see with it.

Yes. If remove the plugin call, the result is the same. That way, this configuration not work.

**Comment by roji** [comment]
Try posting that on the ASP.NET repo - this repo is for EFCore.NamingConventions.

**Comment by adiletelf** [comment]
Would be nice to specify which tables need snake_case naming, for example I may want to use default naming for AspNetCore.Identity tables but snake_case for my own. The reason for it is the strange bug:

1) EFCore.NamingConventions renames columns in AspNetCore.Identity tables like `Id` -> `id` (as expected).
2) I try to use `UserManager<>` class which uses the table `AspNetUsers` to find the user with corresponding id.
3) For some reason instead of using renamed column `id` (lowercase) it tries to search for `Id` (uppercase) and I get Exception: "Column 'AspNetUsers.Id' does not exist".

For now I deleted EFCore.NamingConventions and manually change column names.

**Comment by georg-jung** [comment]
Thanks @roji for the snippet in https://github.com/efcore/EFCore.NamingConventions/issues/2#issuecomment-612651161!

I added 3 more index configurations to have all database object names in snake case (as of identity model v2).

For easy copying and pasting in one's model configuration:

```csharp
private void SetAspNetCoreIdentityDatabaseNamesInSnakeCase(ModelBuilder modelBuilder)
{
    // Adapted from: https://github.com/efcore/EFCore.NamingConventions/issues/2#issuecomment-612651161
    // For identity model v2 as defined in IdentityDbContext.OnModelCreatingVersion2.
    modelBuilder.Entity<IdentityUser>().ToTable("asp_net_users");
    modelBuilder.Entity<IdentityUserToken<string>>().ToTable("asp_net_user_tokens");
    modelBuilder.Entity<IdentityUserLogin<string>>().ToTable("asp_net_user_logins");
    modelBuilder.Entity<IdentityUserClaim<string>>().ToTable("asp_net_user_claims");
    modelBuilder.Entity<IdentityRole>().ToTable("asp_net_roles");
    modelBuilder.Entity<IdentityUserRole<string>>().ToTable("asp_net_user_roles");
    modelBuilder.Entity<IdentityRoleClaim<string>>().ToTable("asp_net_role_claims");

    modelBuilder.Entity<IdentityRole>().HasIndex(r => r.NormalizedName).HasDatabaseName("role_name_index").IsUnique();
    modelBuilder.Entity<IdentityUser>().HasIndex(u => u.NormalizedUserName).HasDatabaseName("user_name_index").IsUnique();
    modelBuilder.Entity<IdentityUser>().HasIndex(u => u.NormalizedEmail).HasDatabaseName("email_index");
}
```
